### ISSUE TITLE
Property access is deoptimized in hooks defined closer to the root of the application

### ISSUE THREAD

**Issue by chasebrewsky** [fastify/fastify#3341]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

3.21.6

### Node.js version

14.7.6

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

11.6

### Description

This bug is related to a similar bug I found recently here: https://github.com/fastify/fastify/issues/3316

After looking into how the framework performs encapsulation I was curious if hooks defined closer to the root level of encapsulation would suffer from the same property access deoptimization since each level of encapsulation defines their own request and reply objects. I found that they do suffer from a similar, if not worse, level of deoptimization then the previous bug the more levels of encapsulation there are.

### Steps to Reproduce

I tested using the following code:

```js
const fastify = require('fastify');

const app = fastify();

let request;

app.addHook('preHandler', (req, reply, done) => {
  const isSame = request ? %HaveSameMap(request, req) : null;

  console.log(isSame);

  request = req;
  done();
});

app.get('/', (req, reply) => {
  reply.send('');
});

app.register((app, opts, done) => {
  app.get('/first', (req, reply) => {
    reply.send('');
  });
  done();
})

app.register((app, opts, done) => {
  app.get('/second', (req, reply) => {
    reply.send('');
  });
  done();
});

app.listen(3000, (err, address) => {
  console.log(`Running on ${address}`);
});
```

Then I did the following fetch requests in my browser:

```js
// Initialize the request object.
fetch('/');

// Test that the request object has the same hidden class as the previous one.
fetch('/');

// Test if the hidden class changes on the first plugin.
fetch('/first');

// Test that the hidden class stays the same.
fetch('/first');

// Test if the hidden class changes on the second plugin.
fetch('/second');

// Test if the hidden class changes on the second plugin.
fetch('/second');
```

And here are the results:

```
$ node --allow-natives-syntax test.js                                                                                                                                                                 
Running on http://127.0.0.1:3000
null
true
false
true
false
true
```

To clarify what each value means:

- null: This was the request object being set to compare on the next request.
- true: This states that the request object uses the same hidden class as the initial request.
- false: This states that the hidden class changes when the first plugin route is used.
- true: This states that the hidden class remains the same when the first plugin route is called again.
- false: This states that the hidden class changes when the second plugin route is used.
- true: This states that the hidden class remains the same when the second plugin route is called again.

### Expected Behavior

I expected the hooks closer to the root of the app to be optimized since those are the ones that are run the most, but the more routes that are defined in varying levels of encapsulation the worst those hooks will perform. The hooks closer to the route will still perform well, but commonly used middleware could potentially suffer from ploymorphism or megamorphism deoptimizations.

**Comment by zekth** [comment]
In the current case the shape difference is only (i think) the encapsulated instance of fastify in the `request`.

In a practical case this would also change properties in `request` / `response` and even the encapsulated `instance`. In my opinion if the request/responses have the same shape within an encapsulated context it's ok.

To have the same shape within all contexts i'm not sure it's doable right now without breaking the whole design.

**Comment by mcollina** [comment]
I'll give it a try and see if my hunch is correct. There are probably some good optimizations that we can apply for this case.

(Note that this is a much better scenario than the previous issue as the shape is determined by ahead-of-time definition vs at-request-time.)

**Comment by chasebrewsky** [comment]
I brought it up because I think it has the potential to become an issue the larger an application becomes due to the fact that its natural to break out bloated pieces of an app into smaller pieces. For example, for an application I made that became pretty large I broke up each API endpoint into the objects that it handles, so it looks similar to this:

```js
app.register(require('./user'));
app.register(require('./alerts'));
app.register(require('./tickets'));
// plus x more different objects.
```

The original file used to hold all these routes became pretty large, so I decided to use the encapsulation in the framework to break them out.

At the root API encapsulation level it has the shared authentication middleware it has to pass through each and every request. Since I broke out all the different API routes into a bunch of different encapsulation levels, then V8 most likely was just like "screw this function in particular" and deoptimized it.

Since this functionality is baked pretty deep into the framework design, it might be better to rely more on conventional guidelines to mitigate accidental de optimization from developers, like I did. Realistically, this really only happens when you break out your endpoints into their own levels of encapsulation since the request and reply objects are made at the level it finds a route match (I think, correct me if I'm wrong). So instead of breaking out my endpoints like I did, then simply changing each route module to be a function that accepts the server object with optional other arguments is much more efficient:

```js
// routes/user.js

module.exports = app => {
  app.get();
  app.post();
};

// routes/index.js

module exports = (app, opts, done) => {
  require('./user')(app);
  // ...etc.
};
```

I don't think it's going to be possible to really fix the issue application wide without potentially breaking encapsulation by moving everything to one shared request and reply object, at least from the parts of the source code that I've seen. I just wanted to bring attention to it in an attempt to prevent others from shooting themself in the foot like I did.

**Comment by mcollina** [comment]
I spent some time investigating the issue. The cost of this megamorphism is roughly 10% throughput, which is significant.

I _think_ it's possible to fix this by completely rewriting the whole decorator logic. I'm not convinced the added complexity is worth it, as decorating a value will break things anyway.

Anyway I have learned my lesson and I would not ship this in v3, but rather try to get it in v4.

---

Side note, you can disable encapsulation by using fastify-plugin and obtain the same result you are after.

**Comment by chasebrewsky** [comment]
> Side note, you can disable encapsulation by using fastify-plugin and obtain the same result you are after.

Oh that's good to know!

**Comment by mcollina** [comment]
Fixed by #3479
